components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+                        return getPublisherDataHolder(resultSet, returnSecrets);
+                    }
+                    return null;
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException(String.format("Error while retrieving subscriber details of id : %s",
+                    subscriberId), e);
+        }
+    }
+
+    /**
+     * Gets all subscriber IDs.
+     *
+     * @param filter search string.
+     * @return list of subscriber IDs.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public List<String> listSubscriberIds(String filter) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+        List<String> subscriberIdList = new ArrayList<>();
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement preparedStmt = new NamedPreparedStatement(connection, GET_SUBSCRIBER_IDS_SQL)) {
+                preparedStmt.setInt(TENANT_ID, tenantId);
+
+                try (ResultSet subscriberIds = preparedStmt.executeQuery()) {
+                    while (subscriberIds.next()) {
+                        subscriberIdList.add(subscriberIds.getString(SUBSCRIBER_ID));
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while retrieving subscriber ids", e);
+        }
+
+        return EntitlementUtil.filterSubscribers(subscriberIdList, filter);
+    }
+
+    /**
+     * Adds a subscriber.
+     *
+     * @param holder publisher data holder.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public void addSubscriber(PublisherDataHolder holder) throws EntitlementException {
+
+        String subscriberId = EntitlementUtil.resolveSubscriberId(holder);
+        if (subscriberId == null) {
+            throw new EntitlementException(ERROR_SUBSCRIBER_ID_NULL);
+        }
+
+        if (isSubscriberExists(subscriberId)) {
+            throw new EntitlementException("Subscriber ID already exists");
+        }
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+        insertSubscriber(subscriberId, holder, propertyDTOs);
+    }
+
+    /**
+     * Updates a subscriber.
+     *
+     * @param holder publisher data holder.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public void updateSubscriber(PublisherDataHolder holder) throws EntitlementException {
+
+        String subscriberId = EntitlementUtil.resolveSubscriberId(holder);
+        if (subscriberId == null) {
+            throw new EntitlementException(ERROR_SUBSCRIBER_ID_NULL);
+        }
+
+        PublisherDataHolder oldHolder;
+        if (isSubscriberExists(subscriberId)) {
+            // TODO: revisit in hybrid impl
+            oldHolder = getSubscriber(subscriberId, false);
```

Comment:
Ref: https://github.com/wso2/carbon-identity-framework/pull/5826/files#r1699832301

Reply:
Oh.. We wont be able to simply change it to true.. This will cause for redundant update operations in [1]. So either
it will have redundant encryption operations (if we keep the value as false) or
it will have redundant update operations (if we keep the value as true)

We should be able to make this simplify and less error prone by building `updatedPropertyDTOs` in this level first and use it to `populateProperties()` and `updateSubscriber()` methods..

[1] https://github.com/wso2/carbon-identity-framework/pull/5826/files#diff-50b8fca7333ced1dc9278978245c1128f4d26e7ed4c0449bd944ea1bf6b25608R293

Reply:
went with building `updatedPropertyDTOs` outside